### ISSUE TITLE
Have search results be relative to currently visible map location

### ISSUE THREAD

**Issue by wmbr** [osmandapp/OsmAnd#14049]
I would strongly prefer if search results were always **centered around the currently visible map portion**, and not around the GPS location.
Furthermore, the little "direction arrows" next to the search results should indicate their positions **relative to the currently viewed location** (up=north, right=east, etc).

I guess this should be a new setting, since other people might prefer the current behavior (search results centered around GPS location and "direction arrows" pointing relative to device orientation.)

**Comment by tradavyvy** [comment]
👍

**Comment by XandrexOSM** [comment]
I think it is better to open two seperate issues. I agree with one and not with the other for instance.

**Comment by Lee-Carre** [comment]
@wmbr

> […] search results […] **centered around the currently visible map portion**, and not around the GPS location. Furthermore, the little "direction arrows" next to the search results should indicate their positions **relative to the currently viewed location** (up=north, right=east, etc).

You can already achieve this;
1. Find an element (PoI or whatever, else tap-and-hold to place a temporary marker) at or near the desired area of interest & select the element.
2. In the UI showing the selected element's details, tap `Actions`
3. In the pop-up UI, tap `Search Nearby`
4. Search for what you're seeking; the results will be relative to the selected element (notice the default background text in the search field)
5. To resume ‘search near my current position’ behaviour, either keep tapping `back` until you reach the main (map) UI again, else (in the Search Nearby selected element UI) tap the (blue) go-to-current-position icon in the top right

> other people might prefer the current behavior (search results centered around GPS location and "direction arrows" pointing relative to device orientation.)

I imagine that this is the common desire, or expected behaviour; users want to know what's nearest them.

Having the described behaviour by default would likely be confusing.

For example, imagine trying to find an uncommon feature. When checking the location (on the map) of each search result, a predictable problem would be the user not knowing that subsequent searches are not centred around his position, but the feature he was viewing. Extra steps would have to be taken to avoid this.
Communicating this all to the user, via UI, would be difficult.

> this should be a new setting

Since the desired functionality already exists, then at most a configuration switch to change the default behaviour.

**Comment by wmbr** [comment]
> You can already achieve this;
> 
That's nice! I didn't know that! 
If you didn't have to first go to the "actions" menu, I would consider that good enough. 

Minor issue: The search results then all just point north. 

> Since the desired functionality already exists, then at most a configuration switch to change the default behaviour.

"Setting" = "configuration switch", no?

**Comment by flooc** [comment]
@Lee-Carre 
Didn't know neither, at least it is feasible without disabling location service. Thanks.

@wmbr 
>  If you didn't have to first go to the "actions" menu, I would consider that good enough.

I agree three taps (one long) is too much. I always want to search around the displayed area, mainly to look for drinking water, camp sites, etc. before getting nearby. And when I look for something around, I can always center the map on my position with a single tap.

Also, I hardly ever check a search result and go back to result list: I usually only use the result list to check my search gives some results nearby before displaying them all on the map.

BTW a quicker way to achieve a "Search nearby" the center of visible map would be perfect. Maybe with a long tap on search button? Or a button similar to the blue go-to-current-position icon in the top right, in the Search Nearby UI: a persistent button (with changing icon) could allow to easily switch between GPS / visible map location.

**Comment by sonora** [comment]
Just a quick addition: I believe it is not necessary to first select a point and then go via _Actions > Search nearby_. The same is achieved by simply **panning the map to any location, and then tap the "Search button"**, as usual. The initial text in the Search field will change to "_Search xx km away_" as a visual indication.

The only thing to remember is that there is some threshold by how much you would have to pan the map from your true position before that occurs. I think that if it is less than 50km, the app assumes you were just panning around looking for it on the map, and then decided to look via "Search" anyway, so within that radius it will default to the true-position-centered search.

PS: And yes, the little direction arrows in the search list
- are blue if relative to your true position. And will indicate the direction relative to North, from where you are.
- are orange if the search centered elsewhere. And will not indicate any direction (always point up). Probably they could indicate the direction from the actual map center (which for a panned map is rather arbitrary). Hence it would likely be rather confusing, even misleading with people not understanding what these directions are relative to.

**Comment by flooc** [comment]
> I think that if it is less than 50km, the app assumes you were just panning around looking for it on the map, and then decided to look via "Search" anyway, so within that radius it will default to the true-position-centered search.

You are right, and the threshold seems to be 50km (with maybe some subtlety on which distance is considered exactly). I understand the idea of automating the choice between GPS/view centering, but 50km is huge when walking or biking. So this is mainly useful for car drivers only.

**Comment by sonora** [comment]
Yes, but also, overall,  the use case of looking for something around where you are is probably a lot more frequent than looking for it a few kilometers away. Hard to say, perhaps the threshold should be lowered to 10 km or so, not sure how that would impact the more frequent use cases.

**Comment by flooc** [comment]
GPS-centered search is more frequent in daily use, but searching some kilometers ahead is very common on bike trips; for example to find a camp site near the area you think you will be able to reach by the end of the day.

I honestly don't care much about this threshold: finding the right value can be nothing but a nightmare and will never match everyone needs for every situation. I'm not saying this heuristic is a bad thing, nor that we can't improve it / use different values for different profiles / make it a parameter. But above all we need an easy way to manually force one behavior or the other when needed. We already have a way to force GPS-centered search with the button in search text area (and the GPS button on the map view); we need one to force view-centered search, even if closer than 50km/10km/whatever value we chose.

**Comment by sonora** [comment]
Yes, and that's currently: `Long-tap there > Actions > Search nearby.`

So I gues the action here, if any, is if someone has an idea how to make a more discoverable UI for this function.

**Comment by flooc** [comment]
And/or, as I proposed earlier:

> BTW a quicker way to achieve a "Search nearby" the center of visible map would be perfect. Maybe with a long tap on search button? Or a button similar to the blue go-to-current-position icon in the top right, in the Search Nearby UI: a persistent button (with changing icon) could allow to easily switch between GPS / visible map location.
